42733: Restore CAS auth functionality

Rillke opened the pull request:
- ~Allow `phpCAS` to `unset` variables~
- do not use deprecated `phpCAS::setDebug()`
- fix call to `phpCAS::client()` by providing ILIAS base URL
- Add more debug logging

https://mantis.ilias.de/view.php?id=42733

Rillke committed:
42733: Restore CAS auth functionality

- Allow phpCAS to unset variables
- do not use deprecated setDebug()
- fix call to phpCAS::client() by providing ILIAS
  base URL
- Add more debug logging

https://mantis.ilias.de/view.php?id=42733

Rillke commented:
> copyright is not as expected in Services/CAS/classes/class.ilAuthProviderCAS.php
> copyright is not as expected in src/HTTP/Wrapper/SuperGlobalDropInReplacement.php

yeah, just tell me what's expected and how to fix it. i suppose i will not change any copyright header.

Rillke committed:
CI obliges me to change copyright header

mjansenDatabay commented on src/HTTP/Wrapper/SuperGlobalDropInReplacement.php:
```diff
@@ -54,6 +54,13 @@ public function offsetSet($offset, $value): void
     public function offsetUnset($offset): void
     {
-        throw new LogicException("Modifying global Request-Array such as \$_GET is not allowed!");
+        if ($offset === 'ticket') {
+            // Allow phpCAS to unset the 'ticket'
+            // they do this in CAS/Client.php#L1070
+            // unset($_GET['ticket']);
+            parent::offsetUnset($offset);
+        } else {
+            throw new LogicException("Modifying global Request-Array such as \$_GET is not allowed!");
+        }
```
Please revert this change and follow the instructions here: https://github.com/ILIAS-eLearning/ILIAS/tree/release_8/src/HTTP#dropinreplacements

Rillke committed:
Revert changes in SuperGlobalDropInReplacement.php

as per Code review:
https://github.com/ILIAS-eLearning/ILIAS/pull/8434#discussion_r1842090761

and try another copyright header as CI job still fails :-/

Rillke committed:
Make copyright checker happy.

mjansenDatabay commented on Services/CAS/classes/class.ilAuthProviderCAS.php:
```diff
@@ -38,29 +67,54 @@ protected function getSettings(): ilCASSettings
                 $this->getSettings()->getPort(),
-                $this->getSettings()->getUri()
+                $this->getSettings()->getUri(),
+                $baseUrl
```
Can't we use `ilUtil::_getHttpPath()` instead?

Rillke replied:
I'm not sure if it is an officially supported feature of ILIAS to work under multiple domain names and if forcing people to a configured domain name (which does not necessarily have to be the one the user is currently on) causes issues anywhere. But it's probably the way with the smallest surface for bad actors.

mjansenDatabay replied:
> if it is an officially supported feature of ILIAS to work under multiple domain names

It's not.

I suggest using the above method to avoid introducing another “truth” regarding the “HTTP path”.

Rillke committed:
Use ilUtil::_getHttpPath()

as per code review.

https://github.com/ILIAS-eLearning/ILIAS/pull/8434#discussion_r1842767942

mjansenDatabay commented on Services/CAS/classes/class.ilAuthProviderCAS.php:
```diff
@@ -38,29 +67,54 @@ protected function getSettings(): ilCASSettings
         try {
-            phpCAS::setDebug(false);
-            phpCAS::setVerbose(true);
+            // If you need a logger, uncomment this, the use statements
+            // and variables above and remove   phpCAS::setLogger(null);
+            // from below.
+            // The logger essentially gives you a trace through the
+            // entire CAS-call-stack.
+            //
+            // $this->logPath = tempnam(sys_get_temp_dir(), 'phpCAS');
+            // $this->logger = new Logger('name');
+            // $handler = new StreamHandler($this->logPath);
+            // $format = "%message%\n";
+            // $formatter = new LineFormatter($format);
+            // $handler->setFormatter($formatter);
+            // $this->logger->pushHandler($handler);
+            // phpCAS::setLogger($this->logger);
```
I suggest injecting the PSR-compliant logger wrapped by the ILIAS logger, e.g.:

```php
$ilias_logger = $DIC->logger()->auth();

phpCAS::setLogger($ilias_logger->getLogger());
```

mjansenDatabay replied:
or even better:

```php
phpCAS::setLogger($this->getLogger()->getLogger());
```

Rillke replied:
@mjansenDatabay I suppose it should be only one `->getLogger()`.

What I noticed is that CAS will still log a lot with log level INFO; user details that might be unwanted in log files, so I will check if phpCAS would somehow allow to be less verbose by default.

```
[0dc17] [2024-11-19 19:27:15.474161] client_auth.INFO: phpCAS::trace:468 47AC .START (2024-11-19 19:27:15) phpCAS-1.6.1 ****************** [CAS.php:468]  
[0dc17] [2024-11-19 19:27:15.474320] client_auth.INFO: phpCAS::traceBegin:352 47AC .=> phpCAS::client('2.0', 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com') [class.ilAuthProviderCAS.php:56]  
[0dc17] [2024-11-19 19:27:15.474460] client_auth.INFO: phpCAS::traceBegin:963 47AC .|    => CAS_Client::__construct('2.0', false, 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com', true, CAS_Session_PhpSession) [CAS.php:369]  
[0dc17] [2024-11-19 19:27:15.474494] client_auth.INFO: phpCAS::trace:982 47AC .|    |    Session is not authenticated [Client.php:982]  
[0dc17] [2024-11-19 19:27:15.474528] client_auth.INFO: phpCAS::traceEnd:1080 47AC .|    <= ''  
[0dc17] [2024-11-19 19:27:15.474545] client_auth.INFO: phpCAS::traceEnd:375 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474562] client_auth.INFO: phpCAS::traceBegin:1702 47AC .=> phpCAS::setNoCasServerValidation() [class.ilAuthProviderCAS.php:59]  
[0dc17] [2024-11-19 19:27:15.474581] client_auth.INFO: phpCAS::trace:1705 47AC .|    You have configured no validation of the legitimacy of the cas server. This is not recommended for production use. [CAS.php:1705]  
[0dc17] [2024-11-19 19:27:15.474596] client_auth.INFO: phpCAS::traceEnd:1707 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474613] client_auth.INFO: phpCAS::traceBegin:1159 47AC .=> phpCAS::forceAuthentication() [class.ilAuthProviderCAS.php:61]  
[0dc17] [2024-11-19 19:27:15.474630] client_auth.INFO: phpCAS::traceBegin:1486 47AC .|    => CAS_Client::forceAuthentication() [CAS.php:1161]  
[0dc17] [2024-11-19 19:27:15.474646] client_auth.INFO: phpCAS::traceBegin:1592 47AC .|    |    => CAS_Client::isAuthenticated() [Client.php:1488]  
...
```

and hundreds of lines more. This is great for log level debug but not for log level info. So please do not yet merge.

Rillke committed:
Use PSR-compliant logger wrapped by ILIAS logger

and pass it to phpCAS::setLogger().

Rillke commented on Services/CAS/classes/class.ilAuthProviderCAS.php:
```diff
@@ -22,14 +26,37 @@
 class ilAuthProviderCAS extends ilAuthProvider
 {
-    private ilCASSettings $settings;
+    // private ilCASSettings $settings;
```
this was unintended.

mjansenDatabay reviewed (changes requested):
As suggested in the inline comments: Please pass the PSR-compliant logger to the CAS-library.

Rillke commented:
> As suggested in the inline comments: Please pass the PSR-compliant logger to the CAS-library.

done

Please don't merge yet. Still looking for a way to make phpCAS not sending trace to log level INFO.

Rillke commented:
@mjansenDatabay it seems phpCAS has hardcoded info-level logging for traces:

See https://github.com/apereo/phpCAS/blob/690d431005e4f88651499458469536201623b057/source/CAS.php#L568

The output looks like:

```
[0dc17] [2024-11-19 19:27:15.474161] client_auth.INFO: phpCAS::trace:468 47AC .START (2024-11-19 19:27:15) phpCAS-1.6.1 ****************** [CAS.php:468]  
[0dc17] [2024-11-19 19:27:15.474320] client_auth.INFO: phpCAS::traceBegin:352 47AC .=> phpCAS::client('2.0', 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com') [class.ilAuthProviderCAS.php:56]  
[0dc17] [2024-11-19 19:27:15.474460] client_auth.INFO: phpCAS::traceBegin:963 47AC .|    => CAS_Client::__construct('2.0', false, 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com', true, CAS_Session_PhpSession) [CAS.php:369]  
[0dc17] [2024-11-19 19:27:15.474494] client_auth.INFO: phpCAS::trace:982 47AC .|    |    Session is not authenticated [Client.php:982]  
[0dc17] [2024-11-19 19:27:15.474528] client_auth.INFO: phpCAS::traceEnd:1080 47AC .|    <= ''  
[0dc17] [2024-11-19 19:27:15.474545] client_auth.INFO: phpCAS::traceEnd:375 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474562] client_auth.INFO: phpCAS::traceBegin:1702 47AC .=> phpCAS::setNoCasServerValidation() [class.ilAuthProviderCAS.php:59]  
[0dc17] [2024-11-19 19:27:15.474581] client_auth.INFO: phpCAS::trace:1705 47AC .|    You have configured no validation of the legitimacy of the cas server. This is not recommended for production use. [CAS.php:1705]  
[0dc17] [2024-11-19 19:27:15.474596] client_auth.INFO: phpCAS::traceEnd:1707 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474613] client_auth.INFO: phpCAS::traceBegin:1159 47AC .=> phpCAS::forceAuthentication() [class.ilAuthProviderCAS.php:61]  
[0dc17] [2024-11-19 19:27:15.474630] client_auth.INFO: phpCAS::traceBegin:1486 47AC .|    => CAS_Client::forceAuthentication() [CAS.php:1161]  
[0dc17] [2024-11-19 19:27:15.474646] client_auth.INFO: phpCAS::traceBegin:1592 47AC .|    |    => CAS_Client::isAuthenticated() [Client.php:1488]  
...
```

We'd better leave `phpCAS::setLogger($this->getLogger());` as a comment?

Rillke committed:
Comment line passing-in ILIAS auth logger to CAS

The phpCAS library will otherwise put a lot into the ILIAS logger
with level INFO, not only limited to technical call information
but will also include information about user accounts trying
to authenticate. For GDPR reasons, someone who has to debug
CAS auth, will have to uncomment that line.

mjansenDatabay commented:
Hi @Rillke ,

thanks a lot for your changes. I agree, the logging is to verbose. We should better leave this as a comment.

After this (Update: You already changed it), I will squash the commits, merge the PR, and cherry-pick the commit to all maintained branches (incl. ILIAS 10).

In general:

What worries me a little is the state of `phpCAS` described here: https://github.com/apereo/phpCAS/tree/690d431005e4f88651499458469536201623b057?tab=readme-ov-file#project-is-unmaintained .
While the library was still accepted for ILIAS 10 (see: https://github.com/ILIAS-eLearning/ILIAS/pull/6765), I doubt that we will be able to do so again for ILIAS 11.
So what we need is an alternative library, preferably within the next three weeks. In December all libraries for ILIAS 11 have to be proposed (again) as a pull request, see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/vendor/README.md#re-evaluating-dependencies-on-major-updates-of-ilias . What we also need is someone who uses/has knowledge in CAS and can push the topic for 11 (together with us). I don't have any “shares” in CAS myself.

If you (or someone else who reads this and needs CAS for authentication in ILIAS) know(s) of alternatives here and want(s) to adapt the code in `ilAuthProviderCAS` even for ILIAS 11 to a newly proposed library via pull request, this would, of course, be enormously helpful for obtaining CAS authentication.

Best regards,
@mjansenDatabay

Rillke commented:
Thanks for the review and the merge. You know, we would pay for such a service, too.

> So what we need is an alternative library, preferably within the next three weeks. In December all libraries for ILIAS 11 have to be proposed (again) as a pull request, see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/vendor/README.md#re-evaluating-dependencies-on-major-updates-of-ilias . What we also need is someone who uses/has knowledge in CAS and can push the topic for 11 (together with us). I don't have any “shares” in CAS myself.

We are currently switching from CAS Server with CAS protocol to Shibboleth Server with CAS protocol (because our IDM team does not want to maintain two SSO software stacks and Shibboleth is more or less mandated by DFN). We will probalbly switch to Shibboleth Server with OIDC soon.

> If you (or someone else who reads this and needs CAS for authentication in ILIAS) know(s) of alternatives here and want(s) to adapt the code in ilAuthProviderCAS even for ILIAS 11 to a newly proposed library via pull request, this would, of course, be enormously helpful for obtaining CAS authentication.

Given that the code was already broken in ILIAS 8 without any issue report in Mantis before mine, I doubt this is widely used. Nonetheless I will look for forums/mailing lists where I could spread the threat of removal in ILIAS 11.

mjansenDatabay commented:
Thx for your contribution @Rillke !
